keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1348 @@
+For MutatingAdmissionPolicy, Object is the alias of the type that the incoming object confirms.
+
+With the object construction syntax, field names are no longer quoted because they are no longer map keys. 
+If any of the object construction violates the defined schema, the expression compilation will error and the user can retrieve the error from the rejection message.
+For list types, i.e. with the OpenAPI type of “array”, the special item field resolves the type of its items.
+In Alpha 1, the CEL environment and its type providers compile the constructed object, but make no effort to check if the field names and types match these of the schemas (i.e. everything is still Dyn). 
+See Construct Type Enforcement in Alpha 2 for future plans.
+
+#### Bindings
+
+Bindings will be almost the same as `ValidatingAdmissionPolicyBinding`, but with
+the following difference:
+
+- No `validationActions` field (unless anyone can think of any useful way to offer a dry-run type option)
+
+#### Parameterization
+Similar to ValidatingAdmissionPolicy, the mutation admission policy can refer to a param, and the param object can be specified per-namespace. 
+We expect to fully reuse existing params handling logic from ValidatingAdmissionPolicy.
+
+#### Reinvocation
+
+The existing reinvocation policy established between webhooks and admission controllers will be extended to also handle admission policies.
+Ref: [the current re-invocation mechanism for webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#reinvocation-policy).
+Admission policies will be reinvoked after admission controllers and before webhooks. 
+
+With mutating admission policies added, the the mutating admission plugin order will become:
+- Mutating admission controllers(e.g. DefaultIngressClass, DefaultStorageClass, etc)
+- Mutating admission policies (introduced within this enhancement and the order will be discussed below)
+- Mutating admission webhooks (ordered lexicographically by webhook name)
+
+To allow mutating admission plugins to observe changes made by other plugins, built-in mutating admission plugins are re-run if a mutating webhook modifies an object, 
+same will apply with mutating policy. The mutating policies are rerun if a mutating webhook modifies an object.
+
+For the running order within mutating admission policies, there are a couple options proposed:
+- option 1(suggested by @deads2k): ordered randomly but keep the same random order while reinvocation.
+  - Pros: 
+    - Encourage user to write oder-independent mutations
+  - Cons: 
+    - The final state of request is not deterministic
+    - The mutation should not have dependencies in between
+- option 2: the lexicographical ordering of the resource names
+  - Pros:
+    - Align with the behavior with mutating webhook
+    - User has a hacky way to enforce the order if it is really needed
```

Comment:
I see it as a con in that it can be used to try to manipulate order (part of the con here is that everyone wants to be last, so it's doesn't actually end up working).

But this does offer determinism, making it easier to validate that some set of mutators works as intended and know it's going to keep working.

Reply:
Updated as a con